packages/rich-text-structure/src/apply-format.js

```diff
@@ -0,0 +1,31 @@
+/**
+ * Applies a given format to a record from the given start to the given end.
+ * If either start or end are omitted, the record's selection will be used.
+ * If at any index there is already a format of the same type present, it will
+ * be removed.
+ *
+ * @param {Object} record     Record to modify.
+ * @param {Object} format     Format to apply.
+ * @param {number} startIndex Start index.
+ * @param {number} endIndex   End index.
+ *
+ * @return {Object} A new record with the format applied.
+ */
+export function applyFormat(
+	{ formats, text, start, end },
+	format,
+	startIndex = start,
+	endIndex = end
+) {
+	for ( let i = startIndex; i < endIndex; i++ ) {
+		if ( formats[ i ] ) {
+			const newFormats = formats[ i ].filter( ( { type } ) => type !== format.type );
+			newFormats.push( format );
```

Comment:
further, since it appears like this code has been written to optimize for speed, we don't need to clone the existing array of formats or run it through the filter.

I'd suggest that _either_ we could write for clarity and simplify this function by removing the `if ( formats[ i ] )` case and maybe use a `map` instead of the loop _or_ we remove this cloning too

```js
if ( formats[ i ] ) {
  const existingFormat = formats[ i ].findIndex( ( { type } ) => type === format.type );
  if ( -1 === existingFormat ) {
    formats[ i ].push( format );
  } else {
    formats[ i ][ existingFormat ] = format;
  }
}
```

Reply:
I'lll have a closer look later, but in this one speed is not important. Speed _is_ important in the `create`, `to-dom` and `to-string` files though. I don't care so much about the manipulation helpers.

Reply:
Oh, here `.map` wouldn't work btw, it will only loop over set array indices.

Reply:
hm…this is from MDN

> Due to the algorithm defined in the specification if the array which map was called upon is sparse, resulting array will also be sparse keeping same indices blank.

are you referring to the fact that `prevFormats` is sparse?

Reply:
Right. So if no format is applied yet, it won't apply any new format because map wouldn't loop over those indices. `[,].map( () => 'test' )` is still `[,]`.